### ISSUE TITLE
Refreshing access token removes refresh token

### ISSUE THREAD

**Issue by mariusznet** [spring-projects/spring-security-oauth#712]
There is a issue with refreshing access token. Apparently `AccessTokenProviderChain` assigns to accessToken the result of method `refreshAccessToken` which is essentially the response from https://tools.ietf.org/html/rfc6749#section-5.1

```java
if (refreshToken != null) {
    accessToken = refreshAccessToken(resource, refreshToken, request);
}
```

Unfortunately returning refresh token in this endpoint is optional, and many providers only returns it when exchanging code for access token or issuing new refresh token. As a result refresh token can be used only once, because next time it will be equal to null.

**Comment by ryanrupp** [comment]
I ran into this issue using Salesforce OAuth with the refresh token process. In the meantime I just extended `AuthorizationCodeAccessTokenProvider` to detect this scenario e.g.:

````
public class FixedRefreshAuthorizationCodeAccessTokenProvider extends AuthorizationCodeAccessTokenProvider {
	    
	    @Override
	    public OAuth2AccessToken refreshAccessToken(OAuth2ProtectedResourceDetails resource,
	            OAuth2RefreshToken refreshToken, AccessTokenRequest request) throws UserRedirectRequiredException,
	            OAuth2AccessDeniedException {
	        OAuth2AccessToken newToken = super.refreshAccessToken(resource, refreshToken, request);
	        // A refresh request typically doesn't return the refresh token, so if it's not there
	        // wrap the token response and set the existing refresh token on it.
	        if (newToken.getRefreshToken() == null) {
	            DefaultOAuth2AccessToken newTokenWithRefreshToken = new DefaultOAuth2AccessToken(newToken);
	            newTokenWithRefreshToken.setRefreshToken(refreshToken);
	            newToken = newTokenWithRefreshToken;
	        }
	        
	        return newToken;
	    }
	}
````

Alternatively this could probably be done in `AccessTokenProviderChain`

**Comment by jgrandja** [comment]
Thanks for the report @mariusznet. The fix has been applied and is now in master.

**Comment by coco1979ka** [comment]
With this implementation one can run into an issue, when the authorization server issues a new refresh_token (which is possible and even an option in Spring OAuth provider - reuseExistingTokens(false)).

So instead of overwriting the refresh token with the old one, it should be checked, whether the response contains a refresh token.

**Comment by dsyer** [comment]
@coco1979ka I noticed that as well. Did you open a new issue?
